tests/kernel/ee/interrupt/dmac_order.cpp

```diff
@@ -0,0 +1,169 @@
+#include <common-ee.h>
+#include <kernel.h>
+#include <ee_regs.h>
+#include <string.h>
+
+#define MAX_DMAC_RESULTS 9
+#define MAX_DMAC_HANDLERS 10
+
+static u32 g_dmacSema = 0;
+static u32 g_dmacHandlers[MAX_DMAC_HANDLERS];
+static u32 g_dmacHandlerCount = 0;
+static u32 g_orderPos = 0;
+static u32 g_order[MAX_DMAC_RESULTS];
+
+static const u32 __attribute__((aligned(16))) g_dummyTransferData[4] = {0, 0, 0, 0};
+
+typedef s32 (*DmacHandlerFunction)(s32);
+
+void appendResult(u32 result) {
+	if(g_orderPos != MAX_DMAC_RESULTS) {
+		iSignalSema(g_dmacSema);
+		g_order[g_orderPos] = result;
+		g_orderPos++;
+	}
+}
+
+template <s32 number>
+s32 continueDmacHandler(s32 channel) {
+	appendResult(number);
+	return 0;
+}
+
+template <s32 number>
+s32 cancelDmacHandler(s32 channel) {
+	appendResult(number);
+	return -1;
+}
+
+void initializeTest() {
+	g_orderPos = 0;
+	g_dmacHandlerCount = 0;
+	{
+		ee_sema_t dmacSemaInfo;
+		memset(&dmacSemaInfo, 0, sizeof(dmacSemaInfo));
+		dmacSemaInfo.init_count = 0;
+		dmacSemaInfo.max_count  = MAX_DMAC_RESULTS;
+		dmacSemaInfo.option     = 0;
+		g_dmacSema = CreateSema(&dmacSemaInfo);
+	}
+}
+
+void addTestDmacHandler(DmacHandlerFunction handler, s32 next) {
+	int handlerId = AddDmacHandler(DMAC_TO_SPR, handler, next);
+	g_dmacHandlers[g_dmacHandlerCount] = handlerId;
+	g_dmacHandlerCount++;
+}
+
+void removeTestDmacHandler(int index) {
```

Comment:
Hmm.  Should `addTestDmacHandler` return an id that is used here?  I feel like the `removeTestDmacHandler(1)` along with calling the functions 1/2/3 with a template is confusing, since they're differently indexed.

-[Unknown]